### ISSUE TITLE
Nginx not starting in awx_web images in kubernetes environment

### ISSUE THREAD

**Issue by innocode-devops** [ansible/awx#1297]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
 - Installer

##### SUMMARY
Nginx is unable to start in k8s environment.

```
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2018/02/19 19:17:22 [emerg] 39#0: mkdir() "/var/lib/nginx/tmp/client_body" failed (13: Permission denied)
2018-02-19 19:17:22,214 INFO exited: nginx (exit status 1; not expected)
2018-02-19 19:17:24,220 INFO spawned: 'nginx' with pid 40
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2018/02/19 19:17:24 [emerg] 40#0: mkdir() "/var/lib/nginx/tmp/client_body" failed (13: Permission denied)
2018-02-19 19:17:24,504 INFO exited: nginx (exit status 1; not expected)
2018-02-19 19:17:27,536 INFO spawned: 'nginx' with pid 45
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2018/02/19 19:17:27 [emerg] 45#0: mkdir() "/var/lib/nginx/tmp/client_body" failed (13: Permission denied)
2018-02-19 19:17:28,039 INFO exited: nginx (exit status 1; not expected)
2018-02-19 19:17:29,091 INFO gave up: nginx entered FATAL state, too many start retries too quickly
...
```
Naturally web interface is not accessible.

##### ENVIRONMENT
* AWX version: latest docker tag (1.0.3.56 at the time of writing)
* AWX install method: kubernetes (GKE specifically)
* Ansible version:  2.4.3

##### STEPS TO REPRODUCE

1. Install to k8s as per INSTALL.md instructions to GKE cluster
2. View awx_web pod logs

##### EXPECTED RESULTS

AWX web interface is accessible via Ingress IP address

##### ACTUAL RESULTS

Ingress responds with internal server error because underlying AWX service does not respond correctly.

##### ADDITIONAL INFORMATION

Just as a consideration, why not move nginx to a separate layer? Make app listen on its own service, put its address as an upstream in nginx.conf, put nginx.conf into configmap.

**Comment by shanemcd** [comment]
I am unable to recreate this. I did run into [another issue](https://github.com/ansible/awx/issues/1305), however once I got past this the deployment worked fine:

```
2018-02-20 19:24:34,731 INFO spawned: 'nginx' with pid 27
2018-02-20 19:24:34,794 INFO spawned: 'uwsgi' with pid 28
2018-02-20 19:24:35,892 INFO success: awx-config-watcher entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2018-02-20 19:24:35,892 INFO success: daphne entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2018-02-20 19:24:35,892 INFO success: nginx entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

![screen shot 2018-02-20 at 2 32 56 pm](https://user-images.githubusercontent.com/773186/36445083-29fe8ade-164b-11e8-912f-4fe652985cd8.png)

What version of Kubernetes are you using on GKE?

**Comment by innocode-devops** [comment]
We're using k8s 1.9.2. Errors in my log show some sort of I/O error, not permission based. I've tried 1.0.2 image with same result. Docker/docker-compose install with same image works fine.

**Comment by matburt** [comment]
can you show the i/o error?

**Comment by tcheronneau** [comment]
Hi,

I've got the same errors so I can get mine : 
`nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2018-02-22 21:11:26,236 INFO exited: nginx (exit status 1; not expected)
2018/02/22 21:11:26 [emerg] 61#0: mkdir() "/var/lib/nginx/tmp/client_body" failed (13: Permission denied)
2018-02-22 21:11:26,602 INFO exited: awx-config-watcher (exit status 0; not expected)
2018-02-22 21:11:29,610 INFO spawned: 'awx-config-watcher' with pid 82
2018-02-22 21:11:29,614 INFO spawned: 'nginx' with pid 83
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2018/02/22 21:11:29 [emerg] 83#0: mkdir() "/var/lib/nginx/tmp/client_body" failed (13: Permission denied)
2018-02-22 21:11:29,842 INFO exited: nginx (exit status 1; not expected)
2018-02-22 21:11:30,319 INFO gave up: nginx entered FATAL state, too many start retries too quickly`

When I'm looking more precisly in the container : 
`
bash-4.2$ ls -l /var/log/
drwxrwxrwx 1 999 nginx   4096 Oct 18 08:09 nginx
`
So the right for this folder seems ok (777 pretty open). 

But for the /var/lib : 
`bash-4.2$ ls -l /var/lib/nginx/
ls: cannot access /var/lib/nginx/tmp: Permission denied
total 0
d????????? ? ? ? ?            ? tmp`

Seems a bit broken. 

Also on k8s 1.9.0.

**Comment by innocode-devops** [comment]
Sorry, I might have misjudged it, it might be RBAC issue, but I haven't got the chance to test it yet, I'll get back as soon as I have something.

**Comment by tcheronneau** [comment]
Hum I'm having hard time to understand how RBAC could cause error in a pod especially since errors are folder not mounted ... Could you explain how you get to this conclusion ? 

Thanks

**Comment by jkhelil** [comment]
I have the same error here using 1.03
   Listening on endpoint tcp:port=8051:interface=127.0.0.1
WSGI app 0 (mountpoint='') ready in 3 seconds on interpreter 0x15c0bb0 pid: 32 (default app)
WSGI app 0 (mountpoint='') ready in 3 seconds on interpreter 0x15c0bb0 pid: 35 (default app)
WSGI app 0 (mountpoint='') ready in 3 seconds on interpreter 0x15c0bb0 pid: 31 (default app)
2018-03-07 14:50:09,108 INFO spawned: 'awx-config-watcher' with pid 86
2018-03-07 14:50:09,113 INFO spawned: 'nginx' with pid 87
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2018/03/07 14:50:09 [emerg] 87#0: mkdir() "/var/lib/nginx/tmp/client_body" failed (13: Permission denied)
2018-03-07 14:50:09,133 INFO exited: nginx (exit status 1; not expected)
2018-03-07 14:50:09,189 INFO exited: awx-config-watcher (exit status 0; not expected)
2018-03-07 14:50:12,195 INFO spawned: 'awx-config-watcher' with pid 88
2018-03-07 14:50:12,197 INFO spawned: 'nginx' with pid 89
nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2018/03/07 14:50:12 [emerg] 89#0: mkdir() "/var/lib/nginx/tmp/client_body" failed (13: Permission denied)
2018-03-07 14:50:12,216 INFO exited: nginx (exit status 1; not expected)
2018-03-07 14:50:12,299 INFO gave up: nginx entered FATAL state, too many start retries too quickly
2018-03-07 14:50:12,299 INFO exited: awx-config-watcher (exit status 0; not expected)
2018-03-07 14:50:13,301 INFO gave up: awx-config-watcher entered FATAL state, too many start retries too quickly

**Comment by shanemcd** [comment]
Are you all running k8s in GKE? If so I'm curious if you see the problem if you create a new cluster with the "Legacy Authorization" dropdown set to "Enabled". Not a long term solution, but we need to narrow down the cause before we can give you a solution.

**Comment by tcheronneau** [comment]
For me it's my own k8s setup on bare metal server.

**Comment by shanemcd** [comment]
Hmmm I just noticed that GKE has now rolled back the default version when creating a new cluster to 1.8.7. 1.9.0 isnt even an option. I'll try with 1.9.3 and report back.

**Comment by shanemcd** [comment]
Confirmed, am seeing the same errors on k8s 1.9.x. 1.8.7 is working fine. Will dig deeper.

**Comment by shanemcd** [comment]
I still don't fully understand what's going on but this will at least get nginx to boot:

```diff
diff --git a/installer/kubernetes/templates/deployment.yml.j2 b/installer/kubernetes/templates/deployment.yml.j2
index d6f0beb19..462c4ccfd 100644
--- a/installer/kubernetes/templates/deployment.yml.j2
+++ b/installer/kubernetes/templates/deployment.yml.j2
@@ -20,6 +20,8 @@ spec:
           volumeMounts:
             - mountPath: /etc/tower
               name: awx-application-config
+            - mountPath: /var/lib/nginx/tmp
+              name: nginx
         - name: awx-celery
           image: {{ awx_task_kubernetes_image }}
           volumeMounts:
@@ -92,6 +94,8 @@ spec:
                 path: settings.py
               - key: secret_key
                 path: SECRET_KEY
+        - name: nginx
+          emptyDir: {}
 ---
 apiVersion: v1
 kind: Service
```

Going to look into this more when I have the time.

**Comment by shanemcd** [comment]
@tcheronneau What is your underlying host OS?

**Comment by tcheronneau** [comment]
It's Ubuntu (Xenial) for all hosts.

**Comment by jkhelil** [comment]
On my side, I am using k8s 1.8.7 on openstack vms.
Ubuntu is used on vms with apparmor enabled
No Pod securityPolicy is set on the cluster(it is new on 1.9)
Looking inside the centos image; it is not enabling selinux,

**Comment by shanemcd** [comment]
Spent a _while_ on this last night. It is AppArmor causing the permission denied errors. I still don't fully understand why only the nginx directories are having this problem. If someone has any insight I'd appreciate the knowledge drop. I will have a PR up shortly.

**Comment by shanemcd** [comment]
Updated images will be on Dockerhub sometime today. We're working through some fallout from a dependency upgrade that landed this morning.

**Comment by shanemcd** [comment]
Please pull the latest installer code and redeploy the latest images from dockerhub. I've verified it works on GKE running 1.9.3. If someone is still having issues let me know and I'll reopen.
